File: track2.3-kt/app/src/main/kotlin/fi/oplab/findyagency/workshop/Issuer.kt
package fi.oplab.findyagency.workshop

import com.sendgrid.*;
import com.sendgrid.helpers.mail.*;
import com.sendgrid.helpers.mail.objects.*;
import org.findy_network.findy_common_kt.*

class Pairwise(id: String) {
  val id: String = id
  var email: String? = null
  var verified: Boolean = false
}

class Issuer(
  connection: Connection,
  credDefId: String
) : Listener {
  val connection = connection
  val pwConnections: MutableMap<String, Pairwise> =
    java.util.Collections.synchronizedMap(mutableMapOf<String, Pairwise>())
  val credDefId: String = credDefId
  val sgClient: SendGrid

  init {
    sgClient = SendGrid(System.getenv("SENDGRID_API_KEY"));
  }

  private suspend fun askForEmail(connectionId: String) {
    // Ask for user email via basic message
    connection.protocolClient.sendMessage(
      connectionId,
      "Please enter your email to get started."
    )
  }

  // Send email via SendGrid API
  private suspend fun sendEmail(email: String, content: String) {
    val mail = Mail(
      Email(System.getenv("SENDGRID_SENDER")),
      "Email verification",
      Email(email),
      Content("text/plain", content)
    )

    println("Sending email ${content} to ${email}")

    val request = Request()
    request.setMethod(Method.POST)
    request.setEndpoint("mail/send")
    request.setBody(mail.build())
    sgClient.api(request)
  }

  fun addInvitation(id: String) {
    pwConnections.put(id, Pairwise(id = id))
  }

  suspend fun setEmailVerified(connectionId: String): Boolean {
    // Skip handling if the verification is already done
    val pwConnection = pwConnections.get(connectionId)
    if (pwConnection == null || pwConnection.email == null || pwConnection.verified) {
      return false
    }

    pwConnection.verified = true

    val attrs = mapOf("email" to pwConnection.email!!)

    println("Offer credential, conn id: ${connectionId}, credDefID: ${credDefId}, attrs: ${attrs}")

    // Send credential offer to the other agent
    connection.protocolClient.sendCredentialOffer(
      pwConnection.id,
      attrs,
      credDefId
    )
    return true
  }

  override suspend fun handleNewConnection(
    notification: Notification,
    status: ProtocolStatus.DIDExchangeStatus
  ) {
    val pwConnection = pwConnections.get(notification.connectionID)
    if (pwConnection == null) {
      // Connection was not for issuing, skip
      return
    }

    if (pwConnection.email == null) {
      askForEmail(notification.connectionID)
    }
  }

  override suspend fun handleBasicMessageDone(
    notification: Notification,
    status: ProtocolStatus.BasicMessageStatus
  ) {
    // Skip handling if message was sent by us or
    // the verification is already done
    val pwConnection = pwConnections.get(notification.connectionID)
    if (status.getSentByMe() || pwConnection == null || pwConnection.email != null) {
      return
    }

    println("Basic message ${status.getContent()} received with id ${notification.protocolID}")

    // Some sanity checking
    val email = status.getContent()
    val emailValid = email.split(' ').size == 1 && email.indexOf('@') >= 0

    if (emailValid) {
      // Valid email, do verification
      pwConnection.email = email

      val content = "Please verify your email by clicking the following link:\n http://localhost:3001/email/${pwConnection.id}"
      // Send verification mail
      sendEmail(pwConnection.email!!, content)

      // Send confirmation via basic message
      connection.protocolClient.sendMessage(
        pwConnection.id,
        "Email is on it's way! Please check your mailbox 📫."
      )
    } else {
      // Email invalid, ask again
      askForEmail(pwConnection.id)
    }
  }

  override suspend fun handleIssueCredentialDone(
    notification: Notification,
    status: ProtocolStatus.IssueCredentialStatus
  ) {
    println("Credential issued, conn id: ${notification.connectionID} with id ${notification.protocolID}")

    pwConnections.remove(notification.connectionID)
  }
}

File: track2.2-go/app/handlers/issuer.go
package handlers

import (
	"context"
	"fmt"
	"log"
	"os"
	"strings"
	"sync"

	"github.com/findy-network/agency-workshop/agent"
	"github.com/findy-network/findy-common-go/agency/client"
	"github.com/findy-network/findy-common-go/agency/client/async"
	agency "github.com/findy-network/findy-common-go/grpc/agency/v1"
	"github.com/lainio/err2"
	"github.com/lainio/err2/try"
	"github.com/sendgrid/sendgrid-go"
	"github.com/sendgrid/sendgrid-go/helpers/mail"
)

type Issuer struct {
	agent.DefaultListener
	conn        client.Conn
	connections sync.Map
	credDefID   string
}

type connection struct {
	id       string
	email    string
	verified bool
}

var (
	sgClient = sendgrid.NewSendClient(os.Getenv("SENDGRID_API_KEY"))
)

func NewIssuer(conn client.Conn, credDefID string) *Issuer {
	return &Issuer{
		conn:      conn,
		credDefID: credDefID,
	}
}

func (i *Issuer) getConnection(id string) *connection {
	if anyConn, ok := i.connections.Load(id); ok {
		if conn, ok := anyConn.(*connection); ok {
			return conn
		}
	}
	return nil
}

func (i *Issuer) AddInvitation(id string) {
	i.connections.Store(id, &connection{id: id})
}

func (i *Issuer) HandleNewConnection(
	notification *agency.Notification,
	status *agency.ProtocolStatus_DIDExchangeStatus,
) {
	defer err2.Catch(func(err error) {
		log.Printf("Error handling new connection: %v", err)
	})

	conn := i.getConnection(notification.ConnectionID)

	if conn == nil {
		// Connection was not for issuing, skip
		return
	}

	if conn.email == "" {
		i.askForEmail(conn.id)
	}
}

func (i *Issuer) HandleBasicMesssageDone(
	notification *agency.Notification,
	status *agency.ProtocolStatus_BasicMessageStatus,
) {
	defer err2.Catch(func(err error) {
		log.Printf("Error handling basic message: %v", err)
	})

	conn := i.getConnection(notification.ConnectionID)

	// Skip handling if
	// 1. Connection was not for issuing
	// 2. Message was sent by us
	// 3. Email has been already asked
	if conn == nil || status.SentByMe || conn.email != "" {
		return
	}

	msg := status.Content
	msgValid := len(strings.Split(msg, " ")) == 1 && strings.Contains(msg, "@")

	log.Printf("Basic message %s with protocol id %s completed with %s",
		msg, notification.ProtocolID, conn.id)

	if msgValid {
		i.connections.Store(conn.id, &connection{id: conn.id, email: msg})

		// Create simple verification link
		// Note: in real-world we should use some random value instead of the connection id
		content := fmt.Sprintf("Please verify your email by clicking the following link:\n http://localhost:3001/email?value=%s", conn.id)
		i.sendEmail(content, msg)

		// Send confirmation via basic message
		pw := async.NewPairwise(i.conn, conn.id)
		_ = try.To1(pw.BasicMessage(context.TODO(), "Email is on it's way! Please check your mailbox 📫."))

	} else {
		// If email is invalid, ask again
		i.askForEmail(conn.id)
	}
}

func (i *Issuer) HandleIssueCredentialDone(
	notification *agency.Notification,
	status *agency.ProtocolStatus_IssueCredentialStatus,
) {
	conn := i.getConnection(notification.ConnectionID)

	if conn == nil {
		// Connection was not for issuing, skip
		return
	}

	log.Printf(
		"Credential issued to: %s, with id: %s",
		notification.ConnectionID,
		notification.ProtocolID,
	)

	i.connections.Delete(notification.ConnectionID)
}

func (i *Issuer) SetEmailVerified(connectionID string) (err error) {
	defer err2.Handle(&err)

	conn := i.getConnection(connectionID)

	// Skip handling if
	// 1. Connection was not for issuing
	// 2. Email has not been saved
	// 3. Credential has been already issued
	if conn == nil || conn.email == "" || conn.verified {
		return
	}

	i.connections.Store(conn.id, &connection{id: conn.id, email: conn.email, verified: true})

	// Create credential content
	attributes := make([]*agency.Protocol_IssuingAttributes_Attribute, 1)
	attributes[0] = &agency.Protocol_IssuingAttributes_Attribute{
		Name:  "email",
		Value: conn.email,
	}

	log.Printf(
		"Offer credential, conn id: %s, credDefID: %s, attrs: %v",
		conn.id,
		i.credDefID,
		attributes,
	)

	// Send credential offer to the other agent
	pw := async.NewPairwise(i.conn, conn.id)
	res := try.To1(pw.IssueWithAttrs(
		context.TODO(),
		i.credDefID,
		&agency.Protocol_IssuingAttributes{
			Attributes: attributes,
		}),
	)

	log.Printf("Credential offered: %s", res.GetID())
	return nil
}

func (i *Issuer) askForEmail(connectionID string) (err error) {
	defer err2.Handle(&err)

	// Ask for user email via basic message
	pw := async.NewPairwise(i.conn, connectionID)
	_ = try.To1(pw.BasicMessage(context.TODO(), "Please enter your email to get started."))

	return err
}

func (i *Issuer) sendEmail(content, email string) (err error) {
	defer err2.Handle(&err)

	from := mail.NewEmail("Issuer example", os.Getenv("SENDGRID_SENDER"))
	subject := "Email verification"
	to := mail.NewEmail(email, email) // Change to your recipient
	message := mail.NewSingleEmail(from, subject, to, content, content)

	log.Printf("Sending email %s to %s", content, email)
	_ = try.To1(sgClient.Send(message))

	return err
}
